enterprise/server/cmd/ci_runner/main.go

```diff
@@ -1728,24 +1755,48 @@ func (ws *workspace) fetchRefs(ctx context.Context) error {
 // checkoutRef checks out a reference that the rest of the remote run should run off
 func (ws *workspace) checkoutRef(ctx context.Context) error {
 	checkoutLocalBranchName := *pushedBranch
-	checkoutRef := *commitSHA
-	if checkoutRef == "" {
-		checkoutRef = fmt.Sprintf("%s/%s", gitRemoteName(*pushedRepoURL), *pushedBranch)
+	if checkoutLocalBranchName == "" {
+		checkoutLocalBranchName = *pushedRef
+	}
+
+	checkoutRef := *pushedRef
+	refIsBranch := isBranch(ctx, *pushedRef, *pushedRepoURL, checkoutLocalBranchName)
+	if refIsBranch {
+		checkoutRef = fmt.Sprintf("%s/%s", gitRemoteName(*pushedRepoURL), *pushedRef)
 	}
 
 	// Create the local branch if it doesn't already exist, then update it to point to the checkout ref
 	if _, err := git(ctx, ws.log, "checkout", "--force", "-B", checkoutLocalBranchName, checkoutRef); err != nil {
 		return err
 	}
+
+	if refIsBranch && *pushedBranch == "" {
+		*pushedBranch = *pushedRef
+	} else if !refIsBranch && *commitSHA == "" {
+		*commitSHA = *pushedRef
+	}
+
 	return nil
 }
 
+// If the ref is a branch name, you need to append `gitRemoteName/` in order to check it out.
+// If the ref is a commit sha, you can check the ref out as is, and appending `gitRemoteName`
+// will result in the reference not being found.
+func isBranch(ctx context.Context, ref string, repoURL string, checkoutLocalBranchName string) bool {
+	refAsBranch := fmt.Sprintf("%s/%s", gitRemoteName(repoURL), ref)
+	_, err := git(ctx, io.Discard, "checkout", "--force", "-B", checkoutLocalBranchName, refAsBranch)
```

Comment:
nit: can we use rev-parse or something that doesn't do as much I/O here? it seems a little heavy to do a checkout in order to detect whether something is a branch